lib/Dialect/Triton/Transforms/Combine.cpp

```diff
@@ -101,6 +101,93 @@ class CombineSelectMaskedLoadPattern : public mlir::RewritePattern {
   }
 };
 
+// sum(x[:, :, None] * y[None, :, :], 1)
+// -> dot(x, y)
+class CombineBroadcastMulReducePattern : public mlir::RewritePattern {
+private:
+  static bool isAddF32(const Operation *op) {
+    if (auto addf = dyn_cast_or_null<arith::AddFOp>(op))
+      return addf.getType().getIntOrFloatBitWidth() <= 32;
+    return false;
+  }
+
+  static SmallVector<int> getEqualIndices(ArrayRef<int64_t> x,
+                                          ArrayRef<int64_t> y) {
+    SmallVector<int> res;
+    for (int i = 0; i < x.size(); ++i)
+      if (x[i] == y[i])
+        res.push_back(i);
+    return res;
+  }
+
+public:
+  CombineBroadcastMulReducePattern(mlir::MLIRContext *context)
+      : mlir::RewritePattern(triton::ReduceOp::getOperationName(), 1, context) {
+  }
+
+  mlir::LogicalResult matchAndRewrite(mlir::Operation *op,
+                                      mlir::PatternRewriter &rewriter) const {
+    auto reduceOp = llvm::dyn_cast<triton::ReduceOp>(op);
+    if (!reduceOp)
+      return mlir::failure();
+    // only support reduce with simple addition
+    Region &combineOp = reduceOp.getCombineOp();
+    bool isReduceAdd = combineOp.hasOneBlock() &&
+                       combineOp.front().getOperations().size() == 2 &&
+                       isAddF32(&*combineOp.front().getOperations().begin());
+    if (!isReduceAdd)
+      return mlir::failure();
+    // operand of reduce has to be mul
+    auto mulOp = llvm::dyn_cast_or_null<arith::MulFOp>(
+        reduceOp.getOperand(0).getDefiningOp());
```

Comment:
nit: one trick is to do:
`auto mulOp = reduceOp.getOperand(0).getDefiningOp<arith::MulFOp>()`